### ISSUE TITLE
Pending spawn of a single-user server

### ISSUE THREAD

**Issue by ablekh** [jupyterhub/zero-to-jupyterhub-k8s#1373]
Request to spawn a single-user server results in empty progress page and the following output of the `hub` pod logs. Potentially related to https://discourse.jupyter.org/t/server-becomes-stale-in-a-matter-of-days-even-hours, but maybe not ... Please share your thoughts.

```
[I 2019-08-22 04:07:51.476 JupyterHub log:158] 302 GET / -> /hub (@10.244.1.1) 1.08ms
[I 2019-08-22 04:07:51.920 JupyterHub log:158] 302 GET /hub -> /hub/ (@10.244.1.1) 0.96ms
[I 2019-08-22 04:07:51.985 JupyterHub log:158] 302 GET /hub/ -> /hub/login (@10.244.1.1) 0.96ms
[I 2019-08-22 04:07:52.125 JupyterHub log:158] 200 GET /hub/login (@10.244.1.1) 1.53ms
[I 2019-08-22 04:07:54.999 JupyterHub oauth2:82] OAuth redirect: 'https://<FQDN>/hub/oauth_callback'
[I 2019-08-22 04:07:55.001 JupyterHub log:158] 302 GET /hub/oauth_login?next= -> https://github.com/login/oauth/authorize?response_type=code&redirect_uri=https%3A%2F%2F<FQDN>%2Fhub%2Foauth_callback&client_id=94321b89d3d18586f304&state=[secret] (@10.244.1.1) 2.00ms
[I 2019-08-22 04:07:55.755 JupyterHub base:509] User logged in: ablekh
[I 2019-08-22 04:07:55.757 JupyterHub log:158] 302 GET /hub/oauth_callback?code=[secret]&state=[secret] -> /user/ablekh/ (@10.244.1.1) 340.37ms
[I 2019-08-22 04:07:55.837 JupyterHub log:158] 302 GET /user/ablekh/ -> /hub/user/ablekh/ (@10.244.1.1) 0.76ms
[I 2019-08-22 04:07:55.935 JupyterHub log:158] 302 GET /hub/user/ablekh/ -> /hub/spawn?next=%2Fhub%2Fuser%2Fablekh%2F (ablekh@10.244.1.1) 35.12ms
[I 2019-08-22 04:07:56.032 JupyterHub log:158] 200 GET /hub/spawn?next=%2Fhub%2Fuser%2Fablekh%2F (ablekh@10.244.1.1) 25.44ms
[I 2019-08-22 04:08:15.863 JupyterHub proxy:301] Checking routes
[W 2019-08-22 04:08:17.774 JupyterHub base:724] User ablekh is slow to start (timeout=0)
[I 2019-08-22 04:08:17.806 JupyterHub log:158] 302 POST /hub/spawn?next=%2Fhub%2Fuser%2Fablekh%2F -> /hub/user/ablekh/ (ablekh@10.244.1.1) 218.71ms
[I 2019-08-22 04:08:17.902 JupyterHub base:1022] Pending spawn for ablekh didn't finish in 0.0 seconds
[I 2019-08-22 04:08:17.902 JupyterHub base:1028] ablekh is pending spawn
[I 2019-08-22 04:08:17.903 JupyterHub log:158] 200 GET /hub/user/ablekh/ (ablekh@10.244.1.1) 24.83ms
[I 2019-08-22 04:09:15.874 JupyterHub proxy:301] Checking routes
```

**Comment by ablekh** [comment]
Additional information: after 10-15 min. wait, I decided to try to submit another spawn request, just out of curiosity. And, to my surprise, the progress page "revived", displayed progress correctly and, finally, presented the classic JupyterHub UI. Relevant logs are enclosed below.

Logs after above-mentioned failure, but before the final success (I haven't changed environment at all):
```
[I 2019-08-22 04:10:15.862 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:10:53.689 JupyterHub log:158] 200 GET /hub/admin (ablekh@10.244.1.1) 32.58ms
[I 2019-08-22 04:11:03.020 JupyterHub log:158] 200 GET /hub/spawn/ablekh (ablekh@10.244.1.1) 25.43ms
[W 2019-08-22 04:11:09.191 JupyterHub web:1782] 400 POST /hub/spawn/ablekh (10.244.1.1): ablekh is pending spawn
[W 2019-08-22 04:11:09.193 JupyterHub log:158] 400 POST /hub/spawn/ablekh (ablekh@10.244.1.1) 28.32ms
[I 2019-08-22 04:11:15.862 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:12:15.862 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:13:15.862 JupyterHub proxy:301] Checking routes
[W 2019-08-22 04:13:17.776 JupyterHub user:471] ablekh's server failed to start in 300 seconds, giving up
[E 2019-08-22 04:13:17.854 JupyterHub gen:593] Exception in Future <Task finished coro=<BaseHandler.spawn_single_user.<locals>.finish_user_spawn() done, defined at /usr/local/lib/python3.6/dist-packages/jupyterhub/handlers/base.py:629> exception=TimeoutError('Timeout',)> after timeout
    Traceback (most recent call last):
      File "/usr/local/lib/python3.6/dist-packages/tornado/gen.py", line 589, in error_callback
        future.result()
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/handlers/base.py", line 636, in finish_user_spawn
        await spawn_future
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/user.py", line 489, in spawn
        raise e
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/user.py", line 409, in spawn
        url = await gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
    tornado.util.TimeoutError: Timeout

[E 2019-08-22 04:13:17.856 JupyterHub gen:593] Exception in Future <Task finished coro=<BaseHandler.spawn_single_user.<locals>.finish_user_spawn() done, defined at /usr/local/lib/python3.6/dist-packages/jupyterhub/handlers/base.py:629> exception=TimeoutError('Timeout',)> after timeout
    Traceback (most recent call last):
      File "/usr/local/lib/python3.6/dist-packages/tornado/gen.py", line 589, in error_callback
        future.result()
      File "/usr/local/lib/python3.6/dist-packages/tornado/gen.py", line 589, in error_callback
        future.result()
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/handlers/base.py", line 636, in finish_user_spawn
        await spawn_future
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/user.py", line 489, in spawn
        raise e
      File "/usr/local/lib/python3.6/dist-packages/jupyterhub/user.py", line 409, in spawn
        url = await gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
    tornado.util.TimeoutError: Timeout

[W 2019-08-22 04:13:17.857 JupyterHub users:442] Stream closed while handling /hub/api/users/ablekh/server/progress
[I 2019-08-22 04:13:17.857 JupyterHub log:158] 200 GET /hub/api/users/ablekh/server/progress (ablekh@10.244.1.1) 299279.07ms
[I 2019-08-22 04:14:15.862 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:14:16.086 JupyterHub log:158] 200 GET /hub/api/users (cull-idle@127.0.0.1) 48.96ms
[I 2019-08-22 04:15:15.862 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:16:15.862 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:17:15.863 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:18:15.863 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:19:15.862 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:20:15.862 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:21:15.862 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:22:15.862 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:23:15.862 JupyterHub proxy:301] Checking routes
[E 2019-08-22 04:24:01.805 JupyterHub gen:593] Exception in Future <Future finished exception=ReadTimeoutError("HTTPSConnectionPool(host='<K8s_FQDN>', port=443): Read timed out. (read timeout=None)",)> after timeout
    Traceback (most recent call last):
      File "/usr/local/lib/python3.6/dist-packages/tornado/gen.py", line 589, in error_callback
        future.result()
      File "/usr/local/lib/python3.6/dist-packages/kubespawner/spawner.py", line 1655, in _start
        body=pvc
      File "/usr/lib/python3.6/concurrent/futures/thread.py", line 56, in run
        result = self.fn(*self.args, **self.kwargs)
      File "/usr/local/lib/python3.6/dist-packages/kubespawner/spawner.py", line 1477, in asynchronize
        return method(*args, **kwargs)
      File "/usr/local/lib/python3.6/dist-packages/kubernetes/client/apis/core_v1_api.py", line 6000, in create_namespaced_persistent_volume_claim
        (data) = self.create_namespaced_persistent_volume_claim_with_http_info(namespace, body, **kwargs)
      File "/usr/local/lib/python3.6/dist-packages/kubernetes/client/apis/core_v1_api.py", line 6091, in create_namespaced_persistent_volume_claim_with_http_info
        collection_formats=collection_formats)
      File "/usr/local/lib/python3.6/dist-packages/kubernetes/client/api_client.py", line 321, in call_api
        _return_http_data_only, collection_formats, _preload_content, _request_timeout)
      File "/usr/local/lib/python3.6/dist-packages/kubernetes/client/api_client.py", line 155, in __call_api
        _request_timeout=_request_timeout)
      File "/usr/local/lib/python3.6/dist-packages/kubernetes/client/api_client.py", line 364, in request
        body=body)
      File "/usr/local/lib/python3.6/dist-packages/kubernetes/client/rest.py", line 266, in POST
        body=body)
      File "/usr/local/lib/python3.6/dist-packages/kubernetes/client/rest.py", line 166, in request
        headers=headers)
      File "/usr/local/lib/python3.6/dist-packages/urllib3/request.py", line 72, in request
        **urlopen_kw)
      File "/usr/local/lib/python3.6/dist-packages/urllib3/request.py", line 150, in request_encode_body
        return self.urlopen(method, url, **extra_kw)
      File "/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py", line 323, in urlopen
        response = conn.urlopen(method, u.request_uri, **kw)
      File "/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py", line 638, in urlopen
        _stacktrace=sys.exc_info()[2])
      File "/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py", line 367, in increment
        raise six.reraise(type(error), error, _stacktrace)
      File "/usr/local/lib/python3.6/dist-packages/urllib3/packages/six.py", line 686, in reraise
        raise value
      File "/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py", line 600, in urlopen
        chunked=chunked)
      File "/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py", line 386, in _make_request
        self._raise_timeout(err=e, url=url, timeout_value=read_timeout)
      File "/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py", line 317, in _raise_timeout
        raise ReadTimeoutError(self, url, "Read timed out. (read timeout=%s)" % timeout_value)
    urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='<K8s_FQDN>', port=443): Read timed out. (read timeout=None)

[I 2019-08-22 04:24:15.862 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:24:16.078 JupyterHub log:158] 200 GET /hub/api/users (cull-idle@127.0.0.1) 40.27ms
[I 2019-08-22 04:25:15.862 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:26:15.862 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:27:15.862 JupyterHub proxy:301] Checking routes
```

Logs matching the final success:
```
[I 2019-08-22 04:28:15.863 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:28:50.722 JupyterHub log:158] 200 GET /hub/home (ablekh@10.244.1.1) 75.40ms
[I 2019-08-22 04:28:55.354 JupyterHub log:158] 200 GET /hub/admin (ablekh@10.244.1.1) 25.83ms
[I 2019-08-22 04:28:58.449 JupyterHub log:158] 200 GET /hub/spawn/ablekh (ablekh@10.244.1.1) 23.93ms
[W 2019-08-22 04:29:00.752 JupyterHub base:724] User ablekh is slow to start (timeout=0)
[I 2019-08-22 04:29:00.779 JupyterHub log:158] 302 POST /hub/spawn/ablekh -> /user/ablekh/ (ablekh@10.244.1.1) 179.95ms
[I 2019-08-22 04:29:00.831 JupyterHub spawner:1659] PVC claim-ablekh already exists, so did not create new pvc.
[I 2019-08-22 04:29:00.842 JupyterHub log:158] 302 GET /user/ablekh/ -> /hub/user/ablekh/ (@10.244.1.1) 0.70ms
[I 2019-08-22 04:29:00.958 JupyterHub base:1022] Pending spawn for ablekh didn't finish in 0.0 seconds
[I 2019-08-22 04:29:00.958 JupyterHub base:1028] ablekh is pending spawn
[I 2019-08-22 04:29:15.862 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:29:55.222 JupyterHub log:158] 200 GET /hub/api (@10.244.1.25) 1.37ms
[I 2019-08-22 04:29:56.004 JupyterHub base:638] User ablekh took 55.366 seconds to start
[I 2019-08-22 04:29:56.004 JupyterHub proxy:242] Adding user ablekh to proxy /user/ablekh/ => http://10.244.1.25:8888
[I 2019-08-22 04:29:56.007 JupyterHub users:533] Server ablekh is ready
[I 2019-08-22 04:29:56.008 JupyterHub log:158] 200 GET /hub/api/users/ablekh/server/progress (ablekh@10.244.1.1) 54534.48ms
[I 2019-08-22 04:29:56.300 JupyterHub log:158] 302 GET /hub/user/ablekh/ -> /user/ablekh/?redirects=1 (ablekh@10.244.1.1) 24.67ms
[I 2019-08-22 04:29:56.631 JupyterHub log:158] 302 GET /hub/api/oauth2/authorize?client_id=jupyterhub-user-ablekh&redirect_uri=%2Fuser%2Fablekh%2Foauth_callback&response_type=code&state=[secret] -> /user/ablekh/oauth_callback?code=[secret]&state=[secret] (ablekh@10.244.1.1) 86.95ms
[I 2019-08-22 04:29:56.790 JupyterHub log:158] 200 POST /hub/api/oauth2/token (ablekh@10.244.1.25) 53.11ms
[I 2019-08-22 04:29:56.870 JupyterHub log:158] 200 GET /hub/api/authorizations/token/[secret] (ablekh@10.244.1.25) 54.47ms
[I 2019-08-22 04:30:15.972 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:31:15.963 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:32:15.888 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:33:15.866 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:34:15.866 JupyterHub proxy:301] Checking routes
[I 2019-08-22 04:34:16.061 JupyterHub log:158] 200 GET /hub/api/users (cull-idle@127.0.0.1) 23.57ms
[I 2019-08-22 04:35:15.866 JupyterHub proxy:301] Checking routes
```

**Comment by ablekh** [comment]
Also, it is not clear why would it take 55 seconds to start a single-user server, considering that relevant Docker image is already present on the corresponding node ... Does it match others' experiences?

**Comment by consideRatio** [comment]
What is the status of the spawning user pod during this slow process? Has it actually started properly, if so, what does its logs say? If it hasn't started in normal amount of time, (like 10 seconds), what does kubectl describe say?

**Comment by ablekh** [comment]
@consideRatio I think that the underlying issue is that single-user server pod is not being created (based on `kubectl get pods` command output) even though a spawn request has been made. So, the only relevant available logs are the ones from the hub pod (see above). Additional problem is that the issue is intermittent - after some [unknown] time passes after the failure, another spawn attempt(s) is/are successful. I have difficulty determining what exact factors (and under what conditions) have the impact.

Normal start time (for default image) that I have experienced for this cluster is approximately 1 minute (10 seconds would be great!). In the spawn failure case, `kubectl describe pod` command for hub pod results in an empty `Events` section (the rest looks OK to me; let me know if posting full output will be helpful).

**Comment by consideRatio** [comment]
If it isn't being created, the question is why.
- The k8s api-server didnt get a message for some reason or has become crazy, I'm clueless as to why and it all becomes out of scope of this repo
- The hub is failing to send the request to the k8s api-server. Why? Where does it get stuck in its work to do this? It seemingly kept operating from the logs.

**Comment by ablekh** [comment]
@consideRatio Thank you for sharing your insights. I agree with your logic above. However, regarding _"it seemingly kept operating from the logs"_: in the spawn failure cases, logs are **missing** the following output that might shed some light on what is going on: `JupyterHub spawner:1659] PVC claim-ablekh already exists, so ...` (this is supposed to appear right after the following line: `JupyterHub log:158] 302 GET /user/ablekh/ -> /hub/user/ablekh/`). The missing output points to JupyterHub spawner code, which is not out of scope here. Any thoughts?

**Comment by scivm** [comment]
I am seeing the issue with the request from hub not being sent to kubernetes also without logging showing what is the issue.  In my case, workaround is to go into your hub control panel and "logout".   Then log back in.   Then spawn succeeds (though if you have a pending spawn you have to wait for that to timeout)

startup time is influenced by the amount of time it takes to mount storage.   For example, in Azure AKS with azure disk storage the 55 seconds would be about right since the storage mounts in 40-50 seconds which is according to azure documentation.  If you use azurefile it takes only 2-3 seconds but azurefile has lower IO performance than azuredisk.

**Comment by ablekh** [comment]
@scivm Thank you for sharing your experience. I've moved to other priorities, but nevertheless appreciate the comment.

**Comment by consideRatio** [comment]
@scivm I think this was resolved by fixes in KubeSpawner which are part of a recent release, 0.14.0, which is available in this Helm chart's latest release `0.9.0-n314.hbcc818ca`.

I'll close this issue as resolved for now until someone can reproduce it in something more recent than `0.9.0-n314.hbcc818ca`. (0.9.1 doesn't count, its just a cherry picked commit on top of 0.9.0)
